### PR TITLE
feat(render): basic of tiled rendering

### PR DESCRIPTION
**Pull request by kisvegabor**
Implemented a tiled rendering. Now it works only in FULL refresh mode. It simply divides the screen into `LV_DRAW_SW_DRAW_UNIT_CNT` tiles, adds the `draw_tasks` for each tile, and waits while they are rendered by the draw threads.

It should utilize the CPU cores better as it's 100% certain if there are two independent tiles for two cores there always will be independent areas that the cores can render. 

I've tested it with SDL on Linux and some draw tasks were really faster with 2 or 4 cores, however some weren't faster at all. 

@XuNeo if I recall correctly you have tested the multi-core rendering on embedded hardware. Could you try this update too? 
  

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by kisvegabor** [comment]
Benchmark [here](https://docs.google.com/spreadsheets/d/1paxFHkt84nvULNf6LYR76k0HoXTVsilInXGMAbBNF_Y/edit?usp=drivesdk).

**Review by liamHowatt — COMMENTED** [review]
Amazing!

image
With disp height of 515 and 4 draw units.

Currently there are `ver_res % LV_DRAW_SW_DRAW_UNIT_CNT` rows of pixels undrawn at the bottom of the display.

**Comment by kisvegabor** [comment]
Yeah, these rounding error still needs to be fixed :slightly_smiling_face:

File: src/drivers/sdl/lv_sdl_window.c [hunk 1]
```diff
@@ -132,7 +132,7 @@ lv_display_t * lv_sdl_window_create(int32_t hor_res, int32_t ver_res)
     lv_display_set_render_mode(disp, LV_DISPLAY_RENDER_MODE_DIRECT);
 #endif /*LV_USE_DRAW_SDL == 0*/
     lv_display_add_event_cb(disp, res_chg_event_cb, LV_EVENT_RESOLUTION_CHANGED, NULL);
-
+    lv_display_set_color_format(disp, LV_COLOR_FORMAT_ARGB8888);
```
Comment (liamHowatt) [inline src/drivers/sdl/lv_sdl_window.c]: I had to remove this to use color depth 16 and 24 or else there was a buffer overflow.

File: src/core/lv_refr.c [hunk 2]
```diff
@@ -616,91 +655,106 @@ static void refr_area(const lv_area_t * area_p)
...
+        /* Assume that the the buffer size (can be screen sized or smaller in case of partial mode)
+         * and max tile size are the optimal scenario. From this calculate the ideal tile size
+         * and set the tile count and tile height accordingly.
+         */
+        uint32_t max_tile_cnt = disp_refr->tile_cnt;
+        uint32_t total_buf_size = layer->draw_buf->data_size;
+        uint32_t ideal_tile_size = total_buf_size / max_tile_cnt;
+        uint32_t area_buf_size = lv_area_get_size(area_p) * lv_color_format_get_size(layer->color_format);
+
+        tile_cnt = (area_buf_size + (ideal_tile_size - 1)) / ideal_tile_size; /*Round up*/
```
Comment (liamHowatt) [inline src/core/lv_refr.c:696]: Seems to round down. Should the comment be updated?

**Review by liamHowatt — COMMENTED** [review]
Nice! Working well.

When I set the draw units to `2` in the tests, there are a number of issues. The test environment having issues doesn't necessarily mean there will be issues outside the tests. Adding multiple units to the tests clearly requires some more changes according to #6665

> Now it works only in FULL refresh mode.

It seems like it also does tiling in the other modes too, which you explained in the docs :slightly_smiling_face:

**Comment by XuNeo** [comment]
> @XuNeo if I recall correctly you have tested the multi-core rendering on embedded hardware. Could you try this update too?

CC @bjsylvia @terry0012

File: src/drivers/sdl/lv_sdl_window.c (hunk 1 above)
Comment (kisvegabor) [inline src/drivers/sdl/lv_sdl_window.c]:
```suggestion
```

Comment (kisvegabor) [reply src/drivers/sdl/lv_sdl_window.c]: Ups, debug code.

**Comment by kisvegabor** [comment]
I've just fixed conflicts. We have tested this feature on multiple devices. It seems the speed up is limited by the memory bandwidth, however on a 4 core MPU we could measure double speed with 4 cores. 

I suggest merging this PR now, and adding updates alter if needed. cc @XuNeo @FASTSHIFT @bjsylvia @liamHowatt

**Review by liamHowatt — DISMISSED** [review]
It worked well for me before. Rebase looks good. I'm assuming the changes to parallel debug borders and the one benchmark scene time is meant to stay.

File: src/core/lv_refr.c (hunk 2 above)
Comment (kisvegabor) [reply src/core/lv_refr.c:696]: We have discussed it and agreed that it rounds up :slightly_smiling_face:

**Comment by kisvegabor** [comment]
Oh, the benchmark change wasn't intentional. Just fixed, thank you! 

The parallel debug related changes are intentional, they are fixing a bug.

**Review by liamHowatt — APPROVED** [review]

**Comment by kisvegabor** [comment]
We have tested this feature on some MPUs and MCUsand it worked well. 

@XuNeo @FASTSHIFT could you take a look at it?

**Review by XuNeo — APPROVED** [review]
Nice!

**Pull request merged**